### PR TITLE
Allow conn string via env-var; inteface keys wrapped in quotes; update nvmrc to 8.15.1

### PR DESCRIPTION
**Pull request by rubengmurray**
Hi 👋 

I'd really like to be able to run this on a hook so I've added the ability for it to use env-vars. My thinking is this could be run in a build pipeline / after migrations via an npm script to ensure types are staying as up-to date as possible.

I also noticed the `.nvmrc` is behind the node version required to run it locally (doesn't work with node 6)... and noticed that the interfaces don't create properly if one of the column names starts with a number so I've wrapped them in quotes.

Run like this `node src/cli.js generate -e ENV_VAR_KEY -o dbschema.ts --noHeader`

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -61,14 +65,37 @@ export async function typescriptOfTable(
...
   excludedTables: string[] = [],
   schema: string | null = null,
   options: OptionValues = {},
+  env?: string,
 ): Promise<string> {
-  const db = typeof dbIn === 'string' ? getDatabase(dbIn) : dbIn;
+  const db = getDbFromArgs(dbIn, env);
+
+  if (!db) {
+    throw new Error('Unable to connect to db')
```
Comment (rubengmurray) [inline src/index.ts:97]: This is for typing mainly

File: src/typescript.ts [hunk 2]
```diff
@@ -80,8 +80,8 @@ export function generateTableInterface(
       }
     }
 
-    selectableMembers += `${jsdoc}${columnName}: ${tsType}${possiblyOrNull};\n`;
-    insertableMembers += `${jsdoc}${columnName}${insertablyOptional}: ${tsType}${possiblyOrNull};\n`;
+    selectableMembers += `${jsdoc}'${columnName}': ${tsType}${possiblyOrNull};\n`;
+    insertableMembers += `${jsdoc}'${columnName}'${insertablyOptional}: ${tsType}${possiblyOrNull};\n`;
```
Comment (rubengmurray) [inline src/typescript.ts:84]: column name of 50-15-d doesn't project to an interface correctly, wrapping in quotes takes care of this

**Comment by danvk** [comment]
Hi @rubengmurray 👋 , glad you found pg-to-ts (also curious _how_ you found pg-to-ts!).

Just wanted to give you a heads-up that I merged #5, which cleans up and modernizes some of the tooling in this repo. In particular, all the tests should run and pass both locally and via GitHub Actions.

You changes seem fine -- I'm curious why it's important to put the connection string in an environment variable that's really a config file, though. If you want it in an environment variable, why not run the command as `pg-to-ts $CONNECTION_STRING -o dbschema.ts`? Or if it needs to be in a file, could you just do something like `pg-to-ts $(cat connection-string.txt) -o dbschema.ts`?

**Comment by rubengmurray** [comment]
> Hi @rubengmurray 👋 , glad you found pg-to-ts (also curious _how_ you found pg-to-ts!).

A blog post I can't seem to re-locate at the moment. I'll try and dig it out to share. Could well have been written by you tbh.

> Just wanted to give you a heads-up that I merged #5, which cleans up and modernizes some of the tooling in this repo. In particular, all the tests should run and pass both locally and via GitHub Actions.

Great, lots of 🧹. Looks like `.nvmrc` could still do with updating though.

> You changes seem fine -- I'm curious why it's important to put the connection string in an environment variable that's really a config file, though. If you want it in an environment variable, why not run the command as `pg-to-ts $CONNECTION_STRING -o dbschema.ts`? Or if it needs to be in a file, could you just do something like `pg-to-ts $(cat connection-string.txt) -o dbschema.ts`?

My experience with accessing environment variables in the run-time has been via a `.env`, for which I've always used `dotenv` to pull in the values from. This has been the same running locally, or accessing env-vars in something like `heroku`. Being honest, I'm not as familiar with build pipelines config, so perhaps there is a different practice here.

But based on your comments, my thinking is:

- Putting it in a text file implies I need to store that text file somewhere to access it or I need to grab that text file from somewhere else to run it. 
  - If I'm pulling it in from somewhere, I want it to be from the same place I have it defined to connect to the database, rather than a separate text file.
- Putting it in the command itself once I've grabbed it means I might expose the credentials in the build log when I run the command (if it's accepted that I print the commands i'm running out to the build log - I guess I could obfuscate this somehow).
  - I'd be grabbing from some exposed set of environment variables here... maybe this is where `.env` is not relevant?
- Using $ENV_VAR without `-c` won't work with the current setup as `-c` is demanded. With `-c` I get an `-o` is not provided error.
  - Perhaps provided `-c ${var}` would be a tidier way of passing the data

What's maybe best at this juncture is for me to test this code in a build pipeline a bit more and see how it works. I've only tested this locally against a `.env` and not implemented even the forked script into a build pipeline yet.

Cheers the response.

**Comment by danvk** [comment]
I published a new version of pg-to-ts (4.0.0) that includes your changes to `.nvmrc` and `engines` (thanks!).

While I was updating `yargs` (the argument parser), I noticed that there were two additional ways to invoke `pg-to-ts` that I hadn't been familiar with:

- If you specify `--config path/to/config.json`, it will read flags from a JSON file (or even a JS file, it uses `require`). So you could configure `pg-to-ts` using a file like `{"conn": "postgres://...", "output": "path/to/output.ts"}`.
- You can specify any flag using an environment variable instead. Previously this would have been something like `SCHEMATS_CONN` but I've changed it to `PG_TO_TS_CONN` with 4.0.0.

I've added examples of both of these forms to the `README.md`. Given the latter form, we're already pretty close to being able to do what you want! You just need a way to set those environment variables.

And it looks like `dotenv` [has you covered there](https://github.com/yargs/yargs/issues/622#issuecomment-252512799). You'll just need to invoke `pg-to-ts` using `node`:

    node -r dotenv/config node_modules/bin/pg-to-ts

Let me know if that works for you—from my perspective it's definitely preferable to not have to support a fourth way of passing in parameters. And if you'd like to put up a new PR that just adds the column quoting, I'll happily merge it.

**Comment by rubengmurray** [comment]
Hi @danvk,

I'll try and have a look through this on the weekend, cheers.

**Comment by rubengmurray** [comment]
Hi @danvk 

I'm able to find a solution with a few tweaks... for me the `.js` isn't happy when I try and dynamically retrieve an already defined env-var inside the file... but for completeness, here's where I'm at:

```json // package.json
{ 
  "pg-to-ts": "node pg-to-ts.js && pg-to-ts generate --config pg-to-ts.json"
}
```

```javascript
// pg-to-ts.js
const dotenv = require('dotenv');
const fs = require('fs');
dotenv.config()

const obj = JSON.stringify({
  "conn": process.env.DATABASE_URL,
  "output": "/schemas.ts"
})

fs.writeFileSync('./pg-to-ts.json', obj, { encoding: 'utf-8' });
```

This is good enough to work with tbh, feel free to close this PR.

I'll re-raise if anything else comes up.. and I'll also look to make that slimmed down PR of the string based keys.

Cheers.

**Comment by rubengmurray** [comment]
https://github.com/danvk/pg-to-ts/pull/13/files

**Pull request closed without merging**